static/js/app.js

```diff
@@ -76,58 +85,96 @@ $('#sendML').on('click', function () {
 $("#fileInputControl").on("change", function () {
   console.log("Found " + this.files.length + " files");
   // add each file to documents
-  $(document.body).css('cursor', 'wait');
-  let invalidFiles = [];
   [].forEach.call(this.files, function (file) {
-    // clean up name of string and check if belongs
-    let fileName = file.name.replace(/\s+/g, "_").replace(/[^A-Za-z0-9\.\-\_]/g, '');
-    if (tagModel.docIndex(fileName) === -1) {
-      // check text file
-      if (fileName.match(/.*\.text$|.*\.txt$/g) !== null) {
-        // read, create, and add file
-        let fileReader = new FileReader(file);
-        fileReader.onload = function () {
-          let newDoc = new Doc(fileName, fileReader.result.replace(/[\r\t\f\v\ ]+/g, " "));
-          console.log("Created Doc: " + fileName);
-          addDoc(newDoc);
-        };
-        fileReader.readAsText(file);
+    uploadDocFromFile(file);
+  });
+  this.value = "";
+});
+
+function uploadDocFromFile(file) {
+  // clean up name of string and check if already belongs
+  let fileName = file.name.replace(/\s+/g, "_").replace(/[^A-Za-z0-9\.\-\_]/g, '');
+  if (tagModel.docIndex(fileName) !== -1) {
+    alert("File already uploaded for: '" + fileName + "'\n");
+  }
+  // txt
+  if (fileName.match(/.*\.text$|.*\.txt$/g) !== null) {
+    console.log("Found txt file: '" + fileName + "'");
+    // read, create, and add file
+    let fileReader = new FileReader(file);
+    fileReader.onload = function () {
+      let newDoc = new Doc(fileName, fileReader.result.replace(/[\r\t\f\v\ ]+/g, " "));
+      addDoc(newDoc);
+    };
+    fileReader.readAsText(file);
+  }
+  // json
+  else if (fileName.match(/.*\.json$/g) !== null) {
+    console.log("Found json file: '" + fileName + "'");
+    // read, create, and add file
+    let fileReader = new FileReader(file);
+    fileReader.onload = function () {
+      let newJson = fileReader.result.replace(/[\r\t\f\v\ ]+/g, " ");
+      let errors = loadJsonData(JSON.parse(newJson));
+      if (errors.length > 0) {
+        alert(errors);
+      }
+    };
+    fileReader.readAsText(file);
+  }
+  // zip
+  else if (fileName.match(/.*\.zip$/g) !== null) {
+    console.log("Found zip file: '" + fileName + "'");
+    uploadDocsFromZipFile(file);
+  }
+  // wasn't one of the file types
+  else {
+    alert("File type not supported for: '" + fileName + "'");
+  }
+  // name matches one of the files already uploaded
+}
+
+function uploadDocsFromZipFile(file) {
+  // load zip file
+  JSZip.loadAsync(file).then(function (zip) {
+    // do each file within zip
+    [].forEach.call(Object.keys(zip.files), function (fileName) {
+      if (tagModel.docIndex(fileName) !== -1) {
+        alert("File already uploaded for: '" + fileName + "' in zip");
+        return;
+      }
+      // mac compressed?
+      if (fileName.match(/^__MACOSX/g) !== null) {
+        alert("Ignored __MACOSX compression file: '" + fileName + "'");
+        return;
```

Comment:
on macs, zipping files generates a per-file __MACOSX file..  10 zipped txt files will make 10 alert boxes, 100...  It's gonna get real annoying.  Maybe just silently ignore them and log it to the console?

Reply:
Well, right now, it's a problem with both zip and json files.

Reply:
how do you mean?